// pretty-quick/dist/scms/git.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stageFile = exports.getUnstagedChangedFiles = exports.getChangedFiles = exports.getSinceRevision = exports.detect = exports.name = void 0;

var _findUp = _interopRequireDefault(require("find-up"));

var _execa = _interopRequireDefault(require("execa"));

var _path = require("path");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const name = 'git';
exports.name = name;

const detect = directory => {
  const gitDirectory = _findUp.default.sync('.git', {
    cwd: directory,
    type: 'directory'
  });

  if (gitDirectory) {
    return (0, _path.dirname)(gitDirectory);
  }
};

exports.detect = detect;

const runGit = (directory, args) => _execa.default.sync('git', args, {
  cwd: directory
});

const getLines = execaResult => execaResult.stdout.split('\n');

const getSinceRevision = (directory, {
  staged,
  branch
}) => {
  try {
    const revision = staged ? 'HEAD' : runGit(directory, ['merge-base', 'HEAD', branch || 'master']).stdout.trim();
    return runGit(directory, ['rev-parse', '--short', revision]).stdout.trim();
  } catch (error) {
    if (/HEAD/.test(error.message) || staged && /Needed a single revision/.test(error.message)) {
      return null;
    }

    throw error;
  }
};

exports.getSinceRevision = getSinceRevision;

const getChangedFiles = (directory, revision, staged) => {
  return [...getLines(runGit(directory, ['diff', '--name-only', staged ? '--cached' : null, '--diff-filter=ACMRTUB', revision].filter(Boolean))), ...(staged ? [] : getLines(runGit(directory, ['ls-files', '--others', '--exclude-standard'])))].filter(Boolean);
};

exports.getChangedFiles = getChangedFiles;

const getUnstagedChangedFiles = directory => {
  return getChangedFiles(directory, null, false);
};

exports.getUnstagedChangedFiles = getUnstagedChangedFiles;

const stageFile = (directory, file) => {
  runGit(directory, ['add', file]);
};

exports.stageFile = stageFile;
